Implement "rank achieved" grouping mode

frenzibyte opened the pull request:
- Closes #34495 
- [x] Depends on #34539 for `FilterCriteria.LocalUserId`

RFC. This does not cover online scores not present in the database. Implementation is done as discussed, using a dictionary to achieve O(n) time complexity, as proposed by @peppy. 

Performance is tested through `TestSceneSongSelectGrouping.TestPerformance`, which loads a total of 10k difficulties and 1.2k scores to a temporary realm database (going any higher takes too long to load). I've also confirmed the generated databases are in a temporary folder (`/var/folders/...`) since the database grows up to 7MB to contain all of the imported models.

Code-wise this PR violates rules to achieve the end result, by including realm directly in the grouping class. I think we're at the point where it doesn't really matter if realm lookups are done directly inside the grouping class or encapsulated by a delegate, and if delegates are still favoured then we'll need a delegate which wraps `realm.Run`, and the signature for that won't be pretty (I proposed something similar for collection grouping before and it was faced with disgust).

I've also took the time to tidy up `TestSceneSongSelectGrouping` as it keeps containing more realm/API-based tests.

frenzibyte committed:
Add test and benchmark coverage

frenzibyte committed:
Add "Rank Achieved" grouping mode

peppy commented:
>  (going any higher takes too long to load)

pretty scary since users will hit this number in no time

frenzibyte commented:
Sorry, I meant it will take too long for the test to load all those beatmaps and scores. I've  still waited patiently and tested having 12k scores as opposed to 1.2k scores, and it had almost no effect on the filter time.

peppy committed:
Simplify LINQ usage to appease inspectcode

peppy commented on osu.Game/Beatmaps/BeatmapInfo.cs:
```diff
@@ -157,6 +157,12 @@ public bool Equals(BeatmapInfo? other)
 
         public bool Equals(IBeatmapInfo? other) => other is BeatmapInfo b && Equals(b);
 
+        public override int GetHashCode()
```
What is this..?

peppy commented on osu.Game/Screens/SelectV2/BeatmapCarousel.cs:
```diff
@@ -98,7 +100,7 @@ public BeatmapCarousel()
                 new BeatmapCarouselFilterMatching(() => Criteria!),
                 new BeatmapCarouselFilterSorting(() => Criteria!),
-                grouping = new BeatmapCarouselFilterGrouping(() => Criteria!, () => detachedCollections())
+                grouping = new BeatmapCarouselFilterGrouping(() => Criteria!, () => realm)
```
why was this moved?

frenzibyte replied:
Since I'm including realm directly, I thought there's not really any reason to still have the `detachedCollections` func. If you don't think that's the correct approach and the `realm` func will eventually be removed, then I'll revert it.

peppy commented on osu.Game/Screens/SelectV2/BeatmapCarousel.cs:
```diff
@@ -697,8 +698,6 @@ protected override Drawable GetDrawableForDisplay(CarouselItem item)
         private Sample? spinSample;
         private Sample? randomSelectSample;
 
-        private Func<List<BeatmapCollection>> detachedCollections = null!;
```
Why?

peppy commented on osu.Game/Screens/SelectV2/BeatmapCarouselFilterGrouping.cs:
```diff
@@ -225,18 +233,52 @@ private List<GroupMapping> getGroups(List<CarouselItem> items, FilterCriteria cr
+                    {
+                        var allLocalScores = r.All<ScoreInfo>()
+                                              .Filter($"{nameof(ScoreInfo.User)}.{nameof(RealmUser.OnlineID)} == $0"
+                                                      + $" && {nameof(ScoreInfo.BeatmapInfo)}.{nameof(BeatmapInfo.Hash)} == {nameof(ScoreInfo.BeatmapHash)}"
```
What's this conditional?

frenzibyte replied:
Carried across from `TopLocalRank` (also exists in `LeaderboardManager`), it seems to be for discarding scores of a beatmap version that doesn't match the current one.

peppy replied:
Inline comment mentioning this is the case, please. I can't read brains.

peppy commented on osu.Game/Screens/SelectV2/BeatmapCarouselFilterGrouping.cs:
```diff
@@ -225,18 +233,52 @@ private List<GroupMapping> getGroups(List<CarouselItem> items, FilterCriteria cr
+                                                      + $" && {nameof(ScoreInfo.Ruleset)}.{nameof(RulesetInfo.ShortName)} == $1"
+                                                      + $" && {nameof(ScoreInfo.DeletePending)} == false", criteria.LocalUserId, criteria.Ruleset?.ShortName)
+                                              .OrderByDescending(s => s.TotalScore)
+                                              .ThenBy(s => s.Date);
```
Date has limited precision. Online score id should be preferred whenever available.

frenzibyte replied:
Also carried from https://github.com/ppy/osu/blob/cb354685ca323f4e429e60617077ed5a5175ee3d/osu.Game/Screens/Select/Carousel/TopLocalRank.cs#L78-L77, I can switch it to fall back to ID but we should probably update that code as well? If so, should I bundle it in this PR or in a separate one.

peppy commented on osu.Game/Screens/SelectV2/BeatmapCarouselFilterGrouping.cs:
```diff
@@ -225,18 +233,52 @@ private List<GroupMapping> getGroups(List<CarouselItem> items, FilterCriteria cr
+                        {
+                            Debug.Assert(score.BeatmapInfo != null);
+
+                            if (topRankMapping.ContainsKey(score.BeatmapInfo))
```
Is there a reason to not use hash here?

peppy replied:
The issue I have here is using a *model class* as the key. I don't care if you use ID  or hash, but please use it directly.

frenzibyte commented on osu.Game.Tests/Visual/SongSelectV2/TestSceneSongSelectGrouping.cs:
```diff
@@ -213,17 +199,146 @@ public void TestMyMapsGroupingUpdatesOnUserChange()
+        #region Rank Achieved grouping
+
+        [Test]
+        public void TestRankAchievedGrouping()
```
I've set up this test scene to take away the complications of setting up a local temporary database and API access, I've also stated that in the xmldoc of the test scene class. 

This test scene also helps confirming that the grouping filter is correctly hooked up with those components (especially cases like a user logging out or collections being modified, which isn't necessarily identified by the grouping filter class itself, but rather signaled by `FilterControl` / components of the screen).

frenzibyte committed:
Replace realm access with delegates

frenzibyte committed:
Fix incorrect rank animation when beatmap panels retrieved from pool

Rank animation is played for new panels when scrolling down, showing the
previous rank belonging to the previous beatmap assigned to that
specific panel instance.

peppy committed:
Revert unnecessary changes

peppy committed:
Merge branch 'master' into rank-achieved-grouping

peppy committed:
Use `ExplicitAttribute` instead of manual nunit ignore
